webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+/**
+ * A DataLoader is an object that can be used to batch requests for fetching objects from the server for performance
+ * reasons.
+ */
+abstract class DataLoader<Identifier, Result = unknown> {
+    protected readonly fetchBatch: (identifiers: Identifier[]) => Result;
+    private readonly maxBatchSize: number;
+
+    protected readonly pendingIdentifiers = new Set<Identifier>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Result;
+        maxBatchSize: number;
+    }) {
+        this.fetchBatch = args.fetchBatch;
+        this.maxBatchSize = args.maxBatchSize;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
+        for (const identifier of identifiersToLoad) {
+            if (!identifier) {
+                continue;
+            }
+
+            this.pendingIdentifiers.add(identifier);
+        }
+    }
+
+    /**
+     * startBatch removes an array of identifiers for data to be loaded from pendingIdentifiers and returns it. If
+     * pendingIdentifiers contains more than maxBatchSize identifiers, then only that many are returned, but if it
+     * contains fewer than that, all of the identifiers are returned and pendingIdentifiers is cleared.
+     */
+    protected startBatch(): {identifiers: Identifier[]; moreToLoad: boolean} {
+        let identifiersToLoad;
```

Comment:
I'm going to leave that one as-is since I like the parallels between the various "identifiers" names. I could see `nextIdentifiersBatch`, but then it's even wordier than the current one 😅